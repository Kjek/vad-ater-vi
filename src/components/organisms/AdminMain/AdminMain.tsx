import AdminOption from '@component/molecules/AdminOption/AdminOption';
import CreateRestaurantSettingModal from '@component/molecules/CreateRestaurantSettingModal/CreateRestaurantSettingModal';
import AdminRestaurantSettingsList from '@component/organisms/AdminRestaurantSettingsList/AdminRestaurantSettingsList';
import LoginSection from '@component/organisms/LoginSection/LoginSection';
import { api } from '@util/api';
import { toastError, toastSuccessful } from '@util/toast-utils';
import { signOut, useSession } from 'next-auth/react';
import { useEffect, useRef, useState } from 'react';
import type { Id } from 'react-toastify';
import { toast } from 'react-toastify';

const AdminMain = () => {
  const { data: session } = useSession();
  const toastId = useRef<Id>(0);
  const [currentDebugRestaurantId, setDebugRestaurantId] = useState<string>('');

  const { data: restaurantSettings, refetch: fetchRestaurants } =
    api.admin.getRestaurantSettings.useQuery(undefined);

  const { mutate: reScrape } = api.admin.reScrape.useMutation({
    onSuccess() {
      toastSuccessful(toastId.current);
    },
    onError(error) {
      toastError(toastId.current, error);
    },
    onMutate() {
      toastId.current = toast.loading('Scraping in progress...');
    },
  });

  const { mutate: reScrapeAll } = api.admin.reScrapeAll.useMutation({
    onSuccess() {
      toastSuccessful(toastId.current);
    },
    onError(error) {
      toastError(toastId.current, error);
    },
    onMutate() {
      toastId.current = toast.loading('Scraping in progress...');
    },
  });

  const { mutate: deleteRestaurant } = api.admin.deleteRestaurant.useMutation({
    onSuccess() {
      toastSuccessful(toastId.current);
      void fetchRestaurants();
    },
    onError(error) {
      toastError(toastId.current, error);
    },
    onMutate() {
      toastId.current = toast.loading('Updating in progress...');
    },
  });

  const { mutate: updateSettings } =
    api.admin.updateRestaurantSetting.useMutation({
      onSuccess() {
        toastSuccessful(toastId.current);
        void fetchRestaurants();
      },
      onError(error) {
        toastError(toastId.current, error);
      },
      onMutate() {
        toastId.current = toast.loading('Updating in progress...');
      },
    });

  const { data: debugData, refetch: fetchDebug } =
    api.admin.debugContent.useQuery(
      { restaurantId: currentDebugRestaurantId },
      { enabled: false, refetchOnWindowFocus: false }
    );

  const { mutate: createRestaurantSetting } =
    api.admin.createRestaurantSetting.useMutation({
      onSuccess() {
        toastSuccessful(toastId.current);
        void fetchRestaurants();
      },
      onError(error) {
        toastError(toastId.current, error);
      },
      onMutate() {
        toastId.current = toast.loading('Creating restaurant...');
      },
    });

  useEffect(() => {
    if (currentDebugRestaurantId) {
      void fetchDebug();
    }
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [currentDebugRestaurantId]);

  return (
    <>
      {session?.user ? (
        <div className='flex items-center justify-center'>
          <div className='dark:bg-gray-custom flex h-full w-full flex-col gap-4 bg-white p-4 md:mx-80 md:my-8 md:p-8'>
            <AdminOption
              text='Re-scrape all restaurants'
              buttonValue='Re-scrape all'
              onClick={() => void reScrapeAll()}
            />
            <AdminRestaurantSettingsList
              restaurantSettings={restaurantSettings}
              debugData={debugData}
              setDebugRestaurantId={setDebugRestaurantId}
              deleteRestaurant={deleteRestaurant}
              updateSettings={updateSettings}
              reScrape={reScrape}
            />
            <CreateRestaurantSettingModal
              createRestaurantSetting={createRestaurantSetting}
            />
            <AdminOption buttonValue='Logout' onClick={() => void signOut()} />
          </div>
        </div>
      ) : (
        <LoginSection />
      )}
    </>
  );
};

export default AdminMain;
