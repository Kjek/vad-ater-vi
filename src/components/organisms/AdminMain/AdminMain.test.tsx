import { render } from '@testing-library/react';
import { describe, expect, it, vi } from 'vitest';
import AdminMain from './AdminMain';

vi.mock('next-auth/react', async () => {
  const originalModule = await vi.importActual('next-auth/react');
  const mockSession = {
    expires: new Date(Date.now() + 2 * 86400).toISOString(),
    user: { username: 'admin' },
  };
  return {
    __esModule: true,
    ...originalModule,
    useSession: vi.fn(() => {
      return { data: mockSession, status: 'authenticated' }; // return type is [] in v3 but changed to {} in v4
    }),
  };
});

vi.mock('@util/api', () => ({
  api: {
    admin: {
      debugContent: {
        useQuery: () => ({
          data: undefined,
          refetch: vi.fn(),
        }),
      },
      getRestaurantSettings: {
        useQuery: () => ({
          data: undefined,
          refetch: vi.fn(),
        }),
      },
      createRestaurantSetting: {
        useMutation: () => ({
          mutate: vi.fn(),
        }),
      },
      updateRestaurantSetting: {
        useMutation: () => ({
          mutate: vi.fn(),
        }),
      },
      deleteRestaurant: {
        useMutation: () => ({
          mutate: vi.fn(),
        }),
      },
      reScrape: {
        useMutation: () => ({
          mutate: vi.fn(),
        }),
      },
      reScrapeAll: {
        useMutation: () => ({
          mutate: vi.fn(),
        }),
      },
    },
  },
}));

describe(AdminMain, () => {
  it('renders correctly', () => {
    const { container } = render(<AdminMain />);
    expect(container).toBeInTheDocument();
    expect(container).toBeVisible();
  });

  it('matches snapshot', () => {
    const defaultRender = render(<AdminMain />);
    expect(defaultRender).toMatchSnapshot();
  });
});
