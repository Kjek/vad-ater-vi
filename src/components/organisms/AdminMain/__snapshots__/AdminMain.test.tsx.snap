// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`AdminMain > matches snapshot 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="flex items-center justify-center"
      >
        <div
          class="dark:bg-gray-custom mx-80 my-8 flex h-full w-full flex-col gap-4 bg-white p-8"
        >
          <div
            class="flex w-full gap-4"
          >
            <h4
              class="font-bold text-gray-800 dark:text-gray-300 flex-1 self-center"
            >
              Re-scrape all restaurants
            </h4>
            <input
              class="cursor-pointer rounded-lg border border-gray-300 px-5 py-2.5 text-sm font-medium text-gray-800 hover:bg-gray-200 dark:border-gray-600 dark:text-gray-300 dark:hover:bg-gray-700 dark:hover:text-white"
              type="button"
              value="Re-scrape all"
            />
          </div>
          <input
            class="rounded-lg border border-gray-300 px-5 py-2.5 text-sm font-medium text-gray-800 hover:bg-gray-200 dark:border-gray-600 dark:text-gray-300 dark:hover:bg-gray-700 dark:hover:text-white flex-1 cursor-pointer self-end"
            type="button"
            value="Add restaurant"
          />
          <div
            class="flex w-full gap-4"
          >
            <h4
              class="font-bold text-gray-800 dark:text-gray-300 flex-1 self-center"
            />
            <input
              class="cursor-pointer rounded-lg border border-gray-300 px-5 py-2.5 text-sm font-medium text-gray-800 hover:bg-gray-200 dark:border-gray-600 dark:text-gray-300 dark:hover:bg-gray-700 dark:hover:text-white"
              type="button"
              value="Logout"
            />
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="flex items-center justify-center"
    >
      <div
        class="dark:bg-gray-custom mx-80 my-8 flex h-full w-full flex-col gap-4 bg-white p-8"
      >
        <div
          class="flex w-full gap-4"
        >
          <h4
            class="font-bold text-gray-800 dark:text-gray-300 flex-1 self-center"
          >
            Re-scrape all restaurants
          </h4>
          <input
            class="cursor-pointer rounded-lg border border-gray-300 px-5 py-2.5 text-sm font-medium text-gray-800 hover:bg-gray-200 dark:border-gray-600 dark:text-gray-300 dark:hover:bg-gray-700 dark:hover:text-white"
            type="button"
            value="Re-scrape all"
          />
        </div>
        <input
          class="rounded-lg border border-gray-300 px-5 py-2.5 text-sm font-medium text-gray-800 hover:bg-gray-200 dark:border-gray-600 dark:text-gray-300 dark:hover:bg-gray-700 dark:hover:text-white flex-1 cursor-pointer self-end"
          type="button"
          value="Add restaurant"
        />
        <div
          class="flex w-full gap-4"
        >
          <h4
            class="font-bold text-gray-800 dark:text-gray-300 flex-1 self-center"
          />
          <input
            class="cursor-pointer rounded-lg border border-gray-300 px-5 py-2.5 text-sm font-medium text-gray-800 hover:bg-gray-200 dark:border-gray-600 dark:text-gray-300 dark:hover:bg-gray-700 dark:hover:text-white"
            type="button"
            value="Logout"
          />
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
